### PR TITLE
fasttrack: proposer-duty: configurable proposer delay

### PR DESCRIPTION
**Pull request by iurii-ssv**
This PR implements https://github.com/ssvlabs/ssv/pull/2260 for **main** branch (so we can release proposer-delay feature sooner).

### CODE REVIEW

File: protocol/v2/ssv/runner/proposer.go [hunk 1]
```diff
@@ -55,11 +60,20 @@ func NewProposerRunner(
...
+	proposerDelay time.Duration,
 ) (Runner, error) {
 	if len(share) != 1 {
 		return nil, errors.New("must have one share")
 	}
 
+	// Cap proposerDelay value to make sure we don't miss block proposals, for details
+	// on how this value should be chosen see:
+	// https://github.com/ssvlabs/ssv/blob/main/docs/MEV_CONSIDERATIONS.md#getting-started-with-mev-configuration
+	const maxReasonableProposerDelay = 1650 * time.Millisecond
```
Comment (kchojn) [inline protocol/v2/ssv/runner/proposer.go]: imo its still magic number, we should move it outside the function. It will be easier to maintain, test and debug in the future

Comment (y0sher) [reply protocol/v2/ssv/runner/proposer.go]: since we merged this to stage this way I prefer we have similar code to prevent conflicts, we can get this to a separate task. Indeed I agree it's a magic number..

**Review by y0sher — DISMISSED** [review]

File: protocol/v2/ssv/runner/proposer.go (hunk 1 above)
Comment (iurii-ssv) [reply protocol/v2/ssv/runner/proposer.go]: Yeah it's most important to keep `stage` and `main` closely aligned,

Regarding "magic numbers", I generally agree with what you are saying - that's why I left a comment linking to the doc, but since we aren't really testing it or anything I don't see much value in moving it somewhere further from the only place where it's used

File: protocol/v2/ssv/runner/proposer.go [hunk 2]
```diff
@@ -55,11 +60,20 @@ func NewProposerRunner(
...
+	if proposerDelay > maxReasonableProposerDelay {
+		proposerDelay = maxReasonableProposerDelay
```
Comment (nkryuchkov) [inline protocol/v2/ssv/runner/proposer.go]: Should we add a log that it's overridden to avoid possible confusion? I know it's a backport of another PR, so maybe just creating an issue would be fine

**Review by nkryuchkov — DISMISSED** [review]

File: protocol/v2/ssv/runner/proposer.go (hunk 2 above)
Comment (y0sher) [reply protocol/v2/ssv/runner/proposer.go]: @iurii-ssv I agree with @nkryuchkov here.

Comment (iurii-ssv) [reply protocol/v2/ssv/runner/proposer.go]: Yes, that would be useful,

let me add a commit for this PR, and also a separate 1-liner PR we can merge into **stage** branch as well

Comment (iurii-ssv) [reply protocol/v2/ssv/runner/proposer.go]: Addressed in https://github.com/ssvlabs/ssv/pull/2262/commits/e49b0a92cddddfaae07c37dbe817136f8da1aa5d now,

and corresponding PR into **stage** branch is https://github.com/ssvlabs/ssv/pull/2267

**Review by y0sher — DISMISSED** [review]

File: protocol/v2/ssv/runner/proposer.go [hunk 3]
```diff
@@ -123,6 +146,15 @@ func (r *ProposerRunner) ProcessPreConsensus(ctx context.Context, logger *zap.Lo
 		zap.Uint64s("signers", getPreConsensusSigners(r.GetState(), root)),
 		fields.PreConsensusTime(r.measurements.PreConsensusTime()))
 
+	// Sleep the remaining proposerDelay since slot start, ensuring on-time proposals even if duty began late.
+	slotStartTime := r.BaseRunner.BeaconNetwork.EstimatedTimeAtSlot(duty.Slot)
+	timeIntoSlot := max(time.Since(time.Unix(slotStartTime, 0)), 0)
+	proposerDelayAdjusted := max(r.proposerDelay-timeIntoSlot, 0)
+	time.Sleep(proposerDelayAdjusted)
```
Comment (moshe-blox) [inline protocol/v2/ssv/runner/proposer.go]:
```suggestion
	select {
		case <-time.After(proposerDelayAdjusted):
		case <-ctx.Done():
	}
```

File: cli/operator/node.go [hunk 4]
```diff
@@ -96,6 +97,7 @@ type config struct {
 	KeyStore                     KeyStore                `yaml:"KeyStore"`
 	SSVSigner                    SSVSignerConfig         `yaml:"SSVSigner" env-prefix:"SSV_SIGNER_"`
 	Graffiti                     string                  `yaml:"Graffiti" env:"GRAFFITI" env-description:"Custom graffiti for block proposals" env-default:"ssv.network" `
+	ProposerDelay                time.Duration           `yaml:"ProposerDelay" env:"PROPOSER_DELAY" env-description:"Duration to wait out before requesting Ethereum block to propose if this Operator is proposer-duty Leader (eg. 300ms). See https://github.com/ssvlabs/ssv/blob/main/docs/MEV_CONSIDERATIONS.md#getting-started-with-mev-configuration for detailed instructions on how to use it."`
```
Comment (oleg-ssvlabs) [inline cli/operator/node.go:98]: Considering we fallback to `300ms` when `ProposerDelay` exceeds that value, wouldn’t it be better to validate it on node startup and crash if it’s misconfigured? So the behavior is explicit and avoids silent overrides

Comment (iurii-ssv) [reply cli/operator/node.go:98]: > Considering we fallback to 300ms

When exceeded, we actually fall back to `1.65ms` (which is "max reasonable"), but that's a minor detail

> wouldn’t it be better to validate it on node startup and crash if it’s misconfigured? So the behavior is explicit and avoids silent overrides

That's how I originally implemented it, there are a bunch of options - https://github.com/ssvlabs/ssv/pull/2196#discussion_r2115313125 - each with its own benefits,

I guess the main idea is that we don't want to crash SSV node because of this proposer-delay param, @y0sher @moshe-blox let me know if you think we should actually crash and force Operator to reconfigure

Comment (y0sher) [reply cli/operator/node.go:98]: I think currently it's ok to not crash the node. but mentioning the delay is lowered it ok and necessary.

**Review by nkryuchkov — DISMISSED** [review]

**Review by kchojn — APPROVED** [review]

File: cli/operator/node_test.go [hunk 5]
```diff
@@ -178,3 +180,147 @@ func Test_verifyConfig(t *testing.T) {
...
+				fields := logs[0].ContextMap()
+				require.Contains(t, fields, "proposer_delay")
+				require.Contains(t, fields, "max_reasonable_proposer_delay")
+				require.Equal(t, delay, fields["proposer_delay"])
+				require.Equal(t, 1000*time.Millisecond, fields["max_reasonable_proposer_delay"])
+			})
+		}
+	})
+
+	t.Run("edge case - exactly at limit", func(t *testing.T) {
```
Comment (oleg-ssvlabs) [inline cli/operator/node_test.go]: I don't think this test is really needed.
We have the exact same test case [here](https://github.com/ssvlabs/ssv/pull/2262/files#diff-1702f9d81695026ebc9f32eec97194e5d59547a4aab23d6edd61f61a615a0cdbR191).

The only difference is that there is no warning log assertion. You could add that assertion to the test above.

File: cli/operator/node_test.go [hunk 6]
```diff
@@ -178,3 +180,147 @@ func Test_verifyConfig(t *testing.T) {
...
+		// Should not return error and should not log warning
+		err := validateProposerDelayConfig(logger)
+		require.NoError(t, err)
+
+		// Should not have any warning logs
+		logs := recorded.All()
+		require.Len(t, logs, 0)
+	})
+
+	t.Run("exactly over limit by 1ms", func(t *testing.T) {
```
Comment (oleg-ssvlabs) [inline cli/operator/node_test.go]: same here. The exact same test case exists [here](https://github.com/ssvlabs/ssv/pull/2262/files#diff-1702f9d81695026ebc9f32eec97194e5d59547a4aab23d6edd61f61a615a0cdbR216)

Comment (iurii-ssv) [reply cli/operator/node_test.go]: https://github.com/ssvlabs/ssv/pull/2262/commits/d12a6d19aa39a730d77583c96dfc4fe69179ff67

File: cli/operator/node_test.go (hunk 5 above)
Comment (iurii-ssv) [reply cli/operator/node_test.go]: https://github.com/ssvlabs/ssv/pull/2262/commits/d12a6d19aa39a730d77583c96dfc4fe69179ff67

**Review by oleg-ssvlabs — APPROVED** [review]
🥇

File: protocol/v2/ssv/runner/proposer.go [hunk 7]
```diff
@@ -123,6 +133,19 @@ func (r *ProposerRunner) ProcessPreConsensus(ctx context.Context, logger *zap.Lo
...
 
+	// Sleep the remaining proposerDelay since slot start, ensuring on-time proposals even if duty began late.
+	slotStartTime := r.BaseRunner.BeaconNetwork.EstimatedTimeAtSlot(duty.Slot)
+	timeIntoSlot := max(time.Since(time.Unix(slotStartTime, 0)), 0)
+	proposerDelayAdjusted := max(r.proposerDelay-timeIntoSlot, 0)
+	select {
+	case <-time.After(proposerDelayAdjusted):
+	case <-ctx.Done():
+		return ctx.Err()
+	}
```
Comment (moshe-blox) [inline protocol/v2/ssv/runner/proposer.go]: i think this could be more straightforward to follow:
```suggestion
	slotTime := r.BaseRunner.BeaconNetwork.EstimatedTimeAtSlot(duty.Slot)
	proposeTime := time.Unix(slotTime, 0).Add(r.proposerDelay)
	if timeLeft := time.Until(proposeTime); timeLeft > 0 {
		select {
		case <-time.After(timeLeft):
		case <-ctx.Done():
			return ctx.Err()
		}
	}
```

Comment (iurii-ssv) [reply protocol/v2/ssv/runner/proposer.go]: Applied in https://github.com/ssvlabs/ssv/pull/2262/commits/e48232644823772a067378677c377f66b544530a

**Review by moshe-blox — APPROVED** [review]

**Comment by moshe-blox** [comment]
well done @iurii-ssv

**Review by kchojn — DISMISSED** [review]

File: scripts/spec-alignment/differ.config.yaml [hunk 8]
```diff
@@ -1,4 +1,4 @@
...
...91ee6f543c","be9935e5c2bedaef","60ac8eb415e6c748","119e95426dc4affb","e918003e25abf4cd","44c0efaa4b189434","de6ef5f1aa07b38d","b71a695e14c7e892","c9b7984a5d2604b5","72e9cd10959558de","82f692ac2f5817f3","3e0d3d7ade790448","52595c2a99ba6031","ea6ad19cdc29b230","695cba9190654139","4418aa6010632480","e27b9c1ea1c64357","fe62097e8814c106","1cfd3a3660a62879","ddf404c2905d3b26","8ac6baea1b755066","2bc023c7e062f24b","8c43f241aeac6d75","832fc9a9a2b71d00","34e8b7999dfb14b5","ccc48c575f7ab4eb","9de0e5d104caf802","e46f98bfca35d4a0","db882c2be6efd7d4","6577bf381d78bae2","f66dbe09241b5a4","48884cab13bd3903","fd6796ea6d131c3","dfde38eed6f3305d","9b895440c1743c4e","783942f921da9985","e0c97cefd6558c77","75c75fa4503282","680487ae23f269a1","9b70726386e38170","ef7b63d9848dc185","3fe59bc388b9ef3","dd93140e21e6f919","143c7821cd65bcac","70d29c9e7a4ef797","b26376133e68b3db","31a354962a9bcfa0","50e62cbde34516ee","9273478f6feccd62","eddf20a67e3fddef","5a44c960cc4178dd","5499865be87b7e00","446545b04b35fba","8887c18020ddb307","3adda38e5fadcfb","9f105a95e82012a4","50cea598241b0bc9","6396e2866ef1b30","1c0da45940bfa76f","22b9026cb0221ce0","2011029b33aad319","2ba698695e4fcb6","12dd40a163e41548","e933dbb89bc159ff"]
```
Comment (iurii-ssv) [inline scripts/spec-alignment/differ.config.yaml:1]: The diff is:

```
diff --git a/protocol/v2/ssv/runner/proposer.go@NewProposerRunner b/ssv/proposer.go@NewProposerRunner
index 2ba698695e4fcb6..2222222
--- a/protocol/v2/ssv/runner/proposer.go@NewProposerRunner
+++ b/ssv/proposer.go@NewProposerRunner
@@ -1,38 +1,29 @@
 func NewProposerRunner(
-	domainType DomainType,
 	beaconNetwork BeaconNetwork,
 	share map[phase0.ValidatorIndex]*Share,
 	qbftController *Controller,
-	beacon beacon.BeaconNode,
+	beacon BeaconNode,
 	network Network,
-	signer ekm.BeaconSigner,
-	operatorSigner ssvtypes.OperatorSigner,
-	doppelgangerHandler DoppelgangerProvider,
+	signer BeaconSigner,
+	operatorSigner *OperatorSigner,
 	valCheck ProposedValueCheckF,
 	highestDecidedSlot phase0.Slot,
-	graffiti []byte,
-	proposerDelay time.Duration,
 ) (Runner, error) {
 	if len(share) != 1 {
 		return nil, errors.New("must have one share")
 	}
 	return &ProposerRunner{
 		BaseRunner: &BaseRunner{
 			RunnerRoleType:		RoleProposer,
-			DomainType:		domainType,
 			BeaconNetwork:		beaconNetwork,
 			Share:			share,
 			QBFTController:		qbftController,
 			highestDecidedSlot:	highestDecidedSlot,
 		},
-		beacon:			beacon,
-		network:		network,
-		signer:			signer,
-		operatorSigner:		operatorSigner,
-		doppelgangerHandler:	doppelgangerHandler,
-		valCheck:		valCheck,
-		measurements:		NewMeasurementsStore(),
-		graffiti:		graffiti,
-		proposerDelay:	proposerDelay,
+		beacon:		beacon,
+		network:	network,
+		signer:		signer,
+		operatorSigner:	operatorSigner,
+		valCheck:	valCheck,
 	}, nil
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/runner/proposer.go@ProposerRunner b/ssv/proposer.go@ProposerRunner
index 12dd40a163e41548..2222222
--- a/protocol/v2/ssv/runner/proposer.go@ProposerRunner
+++ b/ssv/proposer.go@ProposerRunner
@@ -1,12 +1,8 @@
 ProposerRunner struct {
 	BaseRunner	*BaseRunner
-	beacon			beacon.BeaconNode
-	network			Network
-	signer			ekm.BeaconSigner
-	operatorSigner		ssvtypes.OperatorSigner
-	doppelgangerHandler	DoppelgangerProvider
-	valCheck		ProposedValueCheckF
-	measurements		measurementsStore
-	graffiti		[]byte
-	proposerDelay	time.Duration
+	beacon		BeaconNode
+	network		Network
+	signer		BeaconSigner
+	operatorSigner	*OperatorSigner
+	valCheck	ProposedValueCheckF
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/runner/proposer.go@ProposerRunner.ProcessPreConsensus b/ssv/proposer.go@ProposerRunner.ProcessPreConsensus
index e933dbb89bc159ff..2222222
--- a/protocol/v2/ssv/runner/proposer.go@ProposerRunner.ProcessPreConsensus
+++ b/ssv/proposer.go@ProposerRunner.ProcessPreConsensus
@@ -1,48 +1,34 @@
 func (r *ProposerRunner) ProcessPreConsensus(signedMsg *PartialSignatureMessages) error {
 	quorum, roots, err := r.BaseRunner.basePreConsensusMsgProcessing(r, signedMsg)
 	if err != nil {
 		return errors.Wrap(err, "failed processing randao message")
 	}
-	duty := r.GetState().StartingDuty.(*ValidatorDuty)
 	if !quorum {
 		return nil
 	}
-	r.measurements.EndPreConsensus()
-	recordPreConsensusDuration(r.measurements.PreConsensusTime(), RoleProposer)
 	root := roots[0]
 	fullSig, err := r.GetState().ReconstructBeaconSig(r.GetState().PreConsensusContainer, root, r.GetShare().ValidatorPubKey[:], r.GetShare().ValidatorIndex)
 	if err != nil {
-		r.BaseRunner.FallBackAndVerifyEachSignature(r.GetState().PreConsensusContainer, root, r.GetShare().Committee, r.GetShare().ValidatorIndex)
+		r.BaseRunner.FallBackAndVerifyEachSignature(r.GetState().PreConsensusContainer, root, r.GetShare().Committee,
+			r.GetShare().ValidatorIndex)
 		return errors.Wrap(err, "got pre-consensus quorum but it has invalid signatures")
 	}
-	slotTime := r.BaseRunner.BeaconNetwork.EstimatedTimeAtSlot(duty.Slot)
-	proposeTime := time.Unix(slotTime, 0).Add(r.proposerDelay)
-	if timeLeft := time.Until(proposeTime); timeLeft > 0 {
-		select {
-		case <-time.After(timeLeft):
-		case <-():
-			return ()
-		}
-	}
-	start := time.Now()
-	duty = r.GetState().StartingDuty.(*ValidatorDuty)
-	obj, ver, err := r.GetBeaconNode().GetBeaconBlock(duty.Slot, r.graffiti, fullSig)
+	duty := r.GetState().StartingDuty.(*ValidatorDuty)
+	obj, ver, err := r.GetBeaconNode().GetBeaconBlock(duty.Slot, r.GetShare().Graffiti, fullSig)
 	if err != nil {
-		return errors.Wrap(err, "failed to get beacon block")
+		return errors.Wrap(err, "failed to get Beacon block")
 	}
-	blockSummary, summarizeErr := summarizeBlock(obj)
 	byts, err := obj.MarshalSSZ()
 	if err != nil {
 		return errors.Wrap(err, "could not marshal beacon block")
 	}
 	input := &ValidatorConsensusData{
 		Duty:		*duty,
 		Version:	ver,
 		DataSSZ:	byts,
 	}
-	r.measurements.StartConsensus()
-	if err := r.BaseRunner.decide(r, duty.Slot, input); err != nil {
+	if err := r.BaseRunner.decide(r, input.Duty.DutySlot(), input); err != nil {
 		return errors.Wrap(err, "can't start new duty runner instance for duty")
 	}
 	return nil
 }
\ No newline at end of file
```

**Review by y0sher — APPROVED** [review]

**Review by oleg-ssvlabs — APPROVED** [review]
